### PR TITLE
Return rejection if `WebSocket` request couldn't be upgraded

### PR DESCRIPTION
**Pull request by davidpdrsn**
Fixes https://github.com/tokio-rs/axum/issues/1132

Rather than panicking we'll return a rejection. Unfortunately hyper doesn't give any details about why the connection couldn't be upgraded.

### CODE REVIEW

**Review by jplatte — COMMENTED** [review]
I suggest somebody who actually knows stuff about WebSockets should review this.

File: axum/src/extract/ws.rs [hunk 1]
```diff
@@ -267,14 +267,15 @@ where
...
+        let sec_websocket_key = req
+            .headers_mut()
+            .remove(header::SEC_WEBSOCKET_KEY)
+            .ok_or(WebSocketKeyHeaderMissing)?;
 
-        let on_upgrade = req.extensions_mut().remove::<OnUpgrade>().unwrap();
+        let on_upgrade = req
+            .extensions_mut()
+            .remove::<OnUpgrade>()
+            .ok_or(MissingUpgradeState)?;
```
Comment (davidpdrsn) [inline axum/src/extract/ws.rs:278]: @seanmonstar How do you recommend axum deals with `OnUpgrade` missing? We got this issue https://github.com/tokio-rs/axum/issues/1132 recently from a user who ran into this.

This code is inspired by warp which just logs if the extension is missing (https://github.com/seanmonstar/warp/blob/master/src/filters/ws.rs#L154).

Is is possible to get more details from hyper about why the upgrade failed?

**Comment by davidpdrsn** [comment]
@jplatte gentle ping (:

**Comment by jplatte** [comment]
I still don't think I'm the right person to review this though 😅

**Comment by davidpdrsn** [comment]
In that case I think I'll just yolo it. Its a pretty safe change and given what sean said it should be correct.

**Pull request merged**